### PR TITLE
Promote v0.1.1: fix numerically broken PyTorch baselines, dual-GPU benchmarks

### PR DESCRIPTION
**Pull request by simonsays1980**
## Summary

Promotes v0.1.1 with corrected, dual-GPU benchmark results, after finding
and fixing a numerical bug in the PyTorch comparison baselines and hardening
the release process around it.

### The PyTorch baselines were numerically broken, and every ratio in this release moved as a result

Five of the six vectorized `torch.compile` baselines (GAE, V-trace,
λ-returns, eligibility-traces, Retrace) computed their decay-weighted scan
via `exp(cumsum(log(decay.clamp(min=1e-38))))`. The `1e-38` clamp -- needed
so `log(0)` at a terminated step doesn't produce `-inf` -- is exactly what
breaks the formula: each termination contributes `log(1e-38) ≈ -87.5` to a
running suffix sum, and float32 underflows (flushing to zero, then
propagating inf/nan through `exp`) past `≈ -103.3`. At this project's ~5%
per-step termination rate, 2-3 terminations anywhere in a window -- expected
within the first couple dozen steps of any row -- is enough. **This was not
an extreme-sequence-length-only failure: it corrupted 90-99% of output
elements at every size this project's release table uses, including the
production regime (num_envs=4096, seq_len=128).**

Every "N× vs. torch.compile" ratio previously published (including in
`README.md` and `benchmarks.md`) was timed against this broken baseline. The
baselines have been rewritten around `parallel_suffix_scan` /
`parallel_prefix_scan` (linear-space `log2(T)`-doubling associative scans,
no log/exp anywhere) and re-verified against a sequential reference at every
shape before being trusted for timing. **All comparisons in this PR improve
in rl-triton's favor relative to the old numbers** -- not because any kernel
got faster, but because the baseline is now doing real work instead of
silently returning garbage.

One caveat carried into `NOTES.md`: the new baseline is the strongest
*correct* baseline built, not necessarily the fastest possible one -- it
pays 6-12 kernel launches per call (vs. 1-2 for the Triton kernel), so
published ratios describe this specific baseline, not an upper bound on
what a correct PyTorch implementation could achieve.

### Other changes in this PR

- **Fixed GPU provenance errors**: several benchmark artifacts and the
  safeguard floor table were mislabeled H100 when actually measured on H200;
  corrected labels and added a genuine `h100_sxm` floor calibration alongside
  `h200_sxm` and `rtx_2000_ada`.
- **Fixed a second `torch.compile`-related bug**: the baseline-validity gate
  itself checked the eager reference instead of the compiled object at 7
  call sites, so it could not have caught the underflow above if the
  compiled object diverged from eager in some other way. Also fixed a
  cross-object compile-state contamination bug in
  `_bench_truncation_headline()` via `torch._dynamo.reset()`.
- **Rebuilt the release lifecycle**: sweeps now stage to
  `docs/benchmark-history/unreleased.md` by default; `--promote <version>`
  moves a reviewed candidate into `benchmarks.md`, archiving the prior
  version. `--release` (a review-bypassing direct-write path) removed.
- **Added `bench_retrace_truncation()`**: Retrace's truncation path had never
  been benchmarked at any shape; it now has a full CONFIGS-grid table like
  the other four truncating algorithms.
- **Warp tuning fix**: corrected over-provisioned `num_warps` at short
  `BLOCK_SIZE` across all 8 scan kernels.
- **PufferLib comparison + PPO end-to-end measurement** added as one-off
  studies (`benchmarks/pufferlib.md`, `benchmarks/ppo_e2e_measurement.py`).
- **Retired session-scoped handoff reports**, folding durable findings into
  `NOTES.md`.
- **Style**: replaced em dashes with `--` across source, tests, and docs.

`benchmarks.md` now reports two GPU sections (NVIDIA H100 80GB HBM3 and
NVIDIA RTX 2000 Ada Generation), both measured against the corrected
baseline and the corrected gate.

## Test plan

- [x] Correctness gate (`assert_correctness` vs. sequential reference)
      passed for every config in both the H100 and RTX 2000 Ada sweeps,
      including the rewritten baselines themselves
- [x] `--parent-sweep` completed cleanly on both cards
- [x] Monotonicity gate violations reviewed -- all in the documented
      overhead-dominated corner (2-16% at sub-0.05ms absolute times), none
      in Retrace or the new Retrace-truncation table
- [x] All `.py` files compile (`python3 -m py_compile`)
- [x] `benchmarks.md` verified to contain both GPU sections, cross-checked
      against the staged candidate before promotion
- [ ] CI (`pytest -m "not perf and not slow"`, safeguard gate on PR)

### CODE REVIEW

File: .github/workflows/gpu-tests.yml [hunk 1]
```diff
@@ -72,40 +98,90 @@ jobs:
...
-      - name: Run full benchmark sweep
+      - name: Install package (editable, no deps)
+        run: pip install --no-deps -e .
+      - name: Assert GPU identity
+        run: |
+          GPU_NAME=$(python -c "import torch; print(torch.cuda.get_device_name(0))")
+          echo "GPU_NAME=$GPU_NAME" >> "$GITHUB_ENV"
+          echo "Detected GPU: $GPU_NAME" >> "$GITHUB_STEP_SUMMARY"
+          ALLOWED="${{ github.event.inputs.allowed_gpus }}"
+          ALLOWED="${ALLOWED:-H100,H200}"
```
Comment (simonsays1980) [inline .github/workflows/gpu-tests.yml]: Needs to add `RTX 2000 Ada`

File: .github/workflows/gpu-tests.yml [hunk 2]
```diff
@@ -5,29 +5,45 @@ on:
...
-    types: [published]        # full benchmark sweep on release tags
   workflow_dispatch:
     inputs:
-      version:
-        description: "Release tag to record in benchmarks.md (e.g. v0.1.0)"
+      allowed_gpus:
+        description: "Comma-separated substrings of acceptable GPU names for release-bench (safety check against another provenance mislabel)"
         required: false
-        default: "unreleased"
+        default: "H100,H200"
```
Comment (simonsays1980) [inline .github/workflows/gpu-tests.yml]: Needs to add `RTX 2000 Ada`

File: benchmarks/pufferlib_ext/build.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
+"""Loads the real PufferLib advantage-calculation CUDA kernel.
+
+PufferLib's advantage kernel does NOT live at `pufferlib/torch_pufferl.py` /
```
Comment (simonsays1980) [inline benchmarks/pufferlib_ext/build.py:3]: This comparison was done to compare to the momentarily most optimized implementation of GAE/V-Trace in open-source.

File: benchmarks/pufferlib_ext/build.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+`puff_advantage_row` (those names do not exist in the published package as of
+PufferLib 3.0.0, verified against the PyPI sdist on 2026-07-23). The actual
+entry points are:
+
+  - Python wrapper: `compute_puff_advantage()` in `pufferlib/pufferl.py`,
+    which dispatches to `torch.ops.pufferlib.compute_puff_advantage`.
+  - CUDA kernel:     `puff_advantage_row_cuda()` in
+    `pufferlib/extensions/cuda/pufferlib.cu` (one CUDA thread per
```
Comment (simonsays1980) [inline benchmarks/pufferlib_ext/build.py:11]: In comparison: `rl-triton` uses an associative $O(logT)$ scan within each thread.

File: benchmarks/benchmark_gae_vs_pufferlib.py [hunk 5]
```diff
@@ -0,0 +1,735 @@
...
+# Massively-parallel-sim regime: short horizon, high env count (Isaac Gym-style).
+SEQ_LENS_SHORT = [8, 16, 32, 64, 128]
+NUM_ENVS_LIST_LARGE = [4096, 8192, 16384, 32768]
+
+N_ITER = 100
+N_TRIALS = 11
+N_AMORTIZED_CALLS = 100
+N_PROFILE_ITER = 20
+
+H100_PEAK_GBPS = 3350.0  # HBM3 datasheet peak, H100 SXM5 80GB
```
Comment (simonsays1980) [inline benchmarks/benchmark_gae_vs_pufferlib.py:160]: Should be repeated for `RTX 2000 Ada`.

File: benchmarks/benchmark_gae_vs_pufferlib.py [hunk 6]
```diff
@@ -0,0 +1,735 @@
...
+
+# Categorical palette (fixed hue order), light-mode -- from the dataviz skill's
+# validated default palette. Color = num_envs (identity); linestyle = impl.
+COLORS = {128: "#2a78d6", 512: "#eb6834", 2048: "#1baf7a", 8192: "#eda100"}
```
Comment (simonsays1980) [inline benchmarks/benchmark_gae_vs_pufferlib.py:164]: Generates a matplotlib crossover plot.

File: benchmarks/ppo_e2e_measurement.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+from a GAE-only change, since GAE finishes well before the optimizer step and
+has no way to affect Adam's cost. That is the signature of the two arms not
+being measured identically (GPU clock state / allocator warmth drifting
+between the two blocks). Fix here: interleave both arms in ONE process, same
+seeds, alternating which arm runs first every iteration, so any slow drift
+affects both arms equally instead of contaminating whichever arm ran second.
+
+Setup: synthetic rollout buffer (no real env -- this measures update-step cost,
+not env-interaction cost), num_envs=4096, seq_len=128, Isaac-Gym-Ant-like MLP
+policy+value net (obs_dim=60, action_dim=8, continuous Gaussian policy).
```
Comment (simonsays1980) [inline benchmarks/ppo_e2e_measurement.py:23]: Best case scenario. For large networks the percentage of GAE should be lower.

File: benchmarks/ppo_e2e_measurement.py [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+    if compiled:
+        net_a_fwd = torch.compile(net_a)
+        net_b_fwd = torch.compile(net_b)
+    else:
+        net_a_fwd, net_b_fwd = net_a, net_b
+
+    opt_a = torch.optim.Adam(net_a.parameters(), lr=3e-4)
+    opt_b = torch.optim.Adam(net_b.parameters(), lr=3e-4)
+
+    gae_baseline = torch.compile(vectorized_gae)
```
Comment (simonsays1980) [inline benchmarks/ppo_e2e_measurement.py:177]: Note, this is a best case scenario. Most open-source code uses PyTorch/numpy loops and implements no truncation handling.

File: benchmarks/pufferlib.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+**Reproducibility note: the numbers below are not reproducible from the committed script
+as-is.** `benchmarks/compare_pufferlib.py` only imports the real pip `pufferlib` package
+(no vendored fallback), which was not installed in the environment these numbers came from;
+they were instead produced by a separate, uncommitted scratch runner against a vendored,
+sha256-pinned JIT build. Re-running `compare_pufferlib.py` as committed will not reproduce
+this file unless pip `pufferlib` is installed with a working prebuilt extension. See the
+method note at the bottom of this file for the full detail.
+
+Capability + design comparison, not a scoreboard. PufferLib's kernels are real, hand-written CUDA: one thread per environment row, sequential O(T) scan within the thread. rl-triton's kernels are Triton: one program per environment row, O(log T) in-SRAM tree reduction via `tl.associative_scan`. Different mechanisms, different tradeoffs -- PufferLib's flat per-thread cost tends to win at very short horizons; rl-triton's parallel scan tends to win as horizon grows.
```
Comment (simonsays1980) [inline benchmarks/pufferlib.md:13]: This is the most interesting finding in this study. Very short horizon is where PufferLib is sgnificantly faster - almost flat.

File: docs/benchmark-history/ppo-e2e-measurement-2026-07-25.md [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+| **total** | **72.7482 ms** | **72.6960 ms** |
+
+GAE share of step: 0.30% (triton arm) / 0.44% (baseline arm). End-to-end speedup: 0.999x.
+
+### hidden=(256, 256), net mode=torch.compile
+
+| stage | Triton GAE | baseline (torch.compile vectorized) GAE |
+|---|---|---|
+| forward | 19.2250 ms (20.3%) | 19.3544 ms (20.3%) |
+| gae | 0.2927 ms (0.3%) | 0.3508 ms (0.4%) |
```
Comment (simonsays1980) [inline docs/benchmark-history/ppo-e2e-measurement-2026-07-25.md:25]: Again, this is best case: in most open-source code PyTorch/numpy loops are used which are ~700-1000x slower.

File: docs/benchmark-history/v0.1.0.md [hunk 11]
```diff
@@ -0,0 +1,115 @@
+# Benchmarks archive: v0.1.0
+
+**Note:** these numbers predate the log-space baseline underflow fix (see NOTES.md's log-space-underflow note) and are not comparable to later releases -- `compile(vec)` here is the old, since-replaced log-space cumsum baseline. Left unaltered below as the historical record of what was actually measured at the time.
```
Comment (simonsays1980) [inline docs/benchmark-history/v0.1.0.md:3]: Note further, the benchmarks are run only on an `RTX 2000 Ada`.

File: src/rl_triton/ops/prefix_sum.py [hunk 12]
```diff
@@ -6,7 +6,15 @@
 
 # Prefix sum carries the fewest registers of any kernel (2 loads, v=1-done, scan).
 # Use the same aggressive warp/stage settings as the other light kernels.
-_WARPS = {512: 8, 1024: 16, 2048: 16, 4096: 16, 8192: 32, 16384: 32}
+# Below 512, BLOCK_SIZE used to fall through .get()'s default (16 warps) --
+# see src/rl_triton/ops/gae.py's _WARPS for the H200 measurement basis (flat
+# device time for num_warps in {1,2,4} at BLOCK_SIZE 8-128, driven by the
+# 32-blocks/SM hard cap on Hopper, register-count-independent at this scale
+# since none of these kernels spill at low warps counts).
+_WARPS = {
```
Comment (simonsays1980) [inline src/rl_triton/ops/prefix_sum.py:17]: Note, defaults to `16`. Write in NOTES.md and maybe mention here.

File: tests/test_gae.py [hunk 13]
```diff
@@ -624,29 +624,37 @@ def vectorized_gae(
...
-
-    Replaces the Python backward loop with a vectorized equivalent using the same
-    log-space suffix-product trick as vectorized_vtrace. Not production-hardened;
-    only used for benchmarking.
+    Fully vectorized GAE -- strong compiled baseline. Thin wrapper around
+    vectorized_gae_with_truncations (truncateds=0), which uses parallel_suffix_scan
+    (linear-space, log2(T)-doubling, no log/exp) instead of a log-space suffix
+    cumsum.
+
+    This function used to compute a log-space suffix product directly
```
Comment (simonsays1980) [inline tests/test_gae.py:632]: This underflow error was the most important finding in this PR. Basically all previous benchmark runs were showing good numbers from PyTorch vectorized baselines although most of the calculations were flawed by this underflow. 

Needs in the future a further investigation of how this could be still tuned in regard of kernel launches. PyTorch will always execute more kernel launches than Triton (also due to `torch.compile`) but the number of launches might be able to be reduced and can increase the speed of the baseline.

File: tests/test_returns.py [hunk 14]
```diff
@@ -693,6 +693,72 @@ def test_vectorized_discounted_returns_with_truncations_correctness():
     torch.testing.assert_close(actual, expected, atol=1e-5, rtol=1e-5)
 
 
+@cuda_only
+@pytest.mark.xfail(strict=False, reason=(
+    "Known, reproducible-in-isolation torch.compile/Inductor miscompilation of "
```
Comment (simonsays1980) [inline tests/test_returns.py:698]: These are failures that occur in PyTorch 2.4.1 and were not reproduced in PyTorch 2.7. Future releases should test explicitly for PyTorch >=2.7.

File: tests/test_returns.py [hunk 15]
```diff
@@ -693,6 +693,72 @@ def test_vectorized_discounted_returns_with_truncations_correctness():
...
+        bootstrap_values = torch.zeros(num_envs, seq_len, device="cuda")
+        return rewards, terminateds, truncateds, bootstrap_values
+
+    # Object A: a torch.compile(...) wrapper exercised at 2 distinct shapes,
+    # with a reset() before each (the existing, shipped Bug-2-addendum
+    # mitigation) -- confirmed in the bisection to make no difference, but
+    # kept here to faithfully match what the real sweep actually does.
+    object_a = torch.compile(vectorized_discounted_returns_with_truncations)
+    for num_envs, seq_len in [(64, 512), (128, 1024)]:
+        torch._dynamo.reset()
```
Comment (simonsays1980) [inline tests/test_returns.py:744]: This is the important line here.

File: tests/bench_release.py [hunk 16]
```diff
@@ -4,34 +4,78 @@
...
 import re
+import subprocess
 import sys
+import tempfile
 from pathlib import Path
 
+# Silence torch.compile/dynamo symbolic-shapes warnings (e.g. "q1 is not in
+# var_ranges, defaulting to unknown range") that otherwise spam stdout on
+# every torch.compile call in this sweep. .runpod/start.sh sets this at the
+# shell level for the CI runner pod; set it here too (setdefault, so an
```
Comment (simonsays1980) [inline tests/bench_release.py:39]: Also important for any other test runs done manually.

File: tests/bench_release.py [hunk 17]
```diff
@@ -4,34 +4,78 @@
...
 import torch
+import torch._dynamo
 import triton
 
+# This sweep compiles ~15 distinct torch.compile(...) wrapper objects, several
+# reused across every shape in CONFIGS (12) plus PRODUCTION_CONFIGS+BOUNDARY_
+# CONFIG (11) -- ~200 (function, shape) compiles total across the process.
+# Raising Dynamo's cache limits well above that avoids one confirmed failure
+# mode (recompiles past the default cache_size_limit=8 silently falling back
+# instead of compiling) even though it turned out NOT to be the cause of the
```
Comment (simonsays1980) [inline tests/bench_release.py:54]: Note, this was the main reason that the more efficient use of scans in the baselines had to be replaced as torch compile was failing (inductor). The slower version with `torch.compile` is still faster than the fatser one without `torch.compile`.

File: tests/bench_release.py [hunk 18]
```diff
@@ -118,35 +192,14 @@ def _ref_retrace(rewards, dones, values, next_q_all, q_values, actions,
...
-    pi_a   = action_probs_target.gather(-1, actions.unsqueeze(-1)).squeeze(-1)
-    c      = lambda_ * torch.clamp(pi_a / action_probs_behavior, max=c_bar)
-    exp_nq = (action_probs_target * next_q_all).sum(dim=-1)
-    u      = rewards + gamma * exp_nq * (1.0 - dones) - q_values
-    c_next          = torch.empty_like(c)
-    c_next[:, :-1]  = c[:, 1:]
-    c_next[:, -1]   = 0.0
-    v = gamma * c_next * (1.0 - dones)
-
-    log_suffix = torch.flip(
```
Comment (simonsays1980) [inline tests/bench_release.py:141]: This double flip was raising indutor errors in `torch.compile`  - it is more efficient as it avoids the loop above.

**Review by simonsays1980 — COMMENTED** [review]
Altogether every looks fine. This is a PR that immeidately turns into a release. One thing to change: add `RTX 2000 Ada` to the devices for the workflow.
